File: sdk/src/test/java/com/cxense/cxensesdk/CxenseSdkTest.java
package com.cxense.cxensesdk;

import com.cxense.cxensesdk.model.ContentUser;
import com.cxense.cxensesdk.model.Event;
import com.cxense.cxensesdk.model.EventRepository;
import com.cxense.cxensesdk.model.UserExternalData;
import com.cxense.cxensesdk.model.UserIdentity;
import com.cxense.cxensesdk.model.WidgetContext;
import com.cxense.cxensesdk.model.WidgetItem;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.junit.Before;
import org.junit.Test;
import org.powermock.core.classloader.annotations.PrepareForTest;

import java.util.Collections;
import java.util.HashSet;
import java.util.concurrent.ScheduledExecutorService;

import retrofit2.Call;
import retrofit2.Callback;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.ArgumentMatchers.isNull;
import static org.mockito.Mockito.verify;
import static org.powermock.api.mockito.PowerMockito.doNothing;
import static org.powermock.api.mockito.PowerMockito.mock;
import static org.powermock.api.mockito.PowerMockito.spy;
import static org.powermock.api.mockito.PowerMockito.when;

/**
 * @author Dmitriy Konopelkin (dmitry.konopelkin@cxense.com) on (2018-10-15).
 */
@PrepareForTest({CxenseConfiguration.class, CxenseSdk.class})
public class CxenseSdkTest extends BaseTest {
    private Call call;
    private LoadCallback callback;
    private ScheduledExecutorService executor;
    private CxenseConfiguration configuration;
    private AdvertisingIdProvider advertisingIdProvider;
    private UserProvider userProvider;
    private EventRepository eventRepository;
    private CxenseSdk cxense;
    private SendTask sendTask;

    @Before
    public void setUp() throws Exception {
        super.setUp();
        call = mock(Call.class);
        callback = mock(LoadCallback.class);
        CxenseApi api = mock(CxenseApi.class, invocation -> call);
        executor = mock(ScheduledExecutorService.class);
        configuration = mock(CxenseConfiguration.class);
        advertisingIdProvider = mock(AdvertisingIdProvider.class);
        userProvider = mock(UserProvider.class);
        eventRepository = mock(EventRepository.class);
        sendTask = mock(SendTask.class);
        cxense = spy(new CxenseSdk(executor, configuration,
                advertisingIdProvider, userProvider, api, mock(ApiErrorParser.class),
                mock(ObjectMapper.class), eventRepository, sendTask));
    }

    private void verifyApiCalled() {
        verify(call).enqueue(any(Callback.class));
    }

    @Test
    public void trackClick() {
        cxense.trackClick(new WidgetItem());
        verifyApiCalled();
    }

    @Test
    public void trackClickUrl() {
        cxense.trackClick("http://example.com");
        verifyApiCalled();
    }

    @Test
    public void loadWidgetRecommendations() {
        doNothing().when(cxense).loadWidgetRecommendations(anyString(), any(WidgetContext.class), isNull(), any(LoadCallback.class));
        WidgetContext widgetContext = new WidgetContext();
        cxense.loadWidgetRecommendations("id", widgetContext, callback);
        verify(cxense).loadWidgetRecommendations("id", widgetContext, null, callback);
    }

    @Test
    public void loadWidgetRecommendationsNullUser() {
        cxense.loadWidgetRecommendations("id", new WidgetContext(), null, callback);
        verify(cxense).getDefaultUser();
        verifyApiCalled();
    }

    @Test
    public void loadWidgetRecommendationsAllArgs() {
        cxense.loadWidgetRecommendations("id", new WidgetContext(), mock(ContentUser.class), callback);
        verifyApiCalled();
    }

    @Test
    public void transform() {
        assertNotNull(cxense.transform(callback));
    }

    @Test
    public void transformFullArgs() {
        Function function = mock(Function.class);
        assertNotNull(cxense.transform(callback, function));
        verify(cxense).transform(any(LoadCallback.class));
    }

    @Test
    public void getUserId() {
        cxense.getUserId();
        verify(userProvider).getUserId();
    }

    @Test
    public void setUserId() {
        String id = "VeryVeryVeryGoodId";
        cxense.setUserId(id);
        verify(userProvider).setUserId(id);
    }

    @Test
    public void getDefaultUserId() {
        cxense.getDefaultUserId();
        verify(advertisingIdProvider).getDefaultUserId();
    }

    @Test
    public void isLimitAdTrackingEnabled() {
        cxense.isLimitAdTrackingEnabled();
        verify(advertisingIdProvider).isLimitAdTrackingEnabled();
    }

    @Test
    public void getConfiguration() {
        assertEquals(configuration, cxense.getConfiguration());
    }

    @Test
    public void getUserSegmentIds() {
        cxense.getUserSegmentIds(Collections.emptyList(), Collections.emptyList(), callback);
        verifyApiCalled();
    }

    @Test
    public void getUserSegmentIdsConsentDisallowed() {
        when(configuration.getConsentOptions()).thenReturn(new HashSet<>(Collections.singletonList(ConsentOption.CONSENT_REQUIRED)));
        cxense.getUserSegmentIds(Collections.emptyList(), Collections.emptyList(), callback);
        verify(configuration).getConsentOptions();
        verify(callback).onSuccess(Collections.emptyList());
    }

    @Test(expected = IllegalArgumentException.class)
    public void getUserSegmentIdsNullIdentities() {
        cxense.getUserSegmentIds(null, Collections.emptyList(), callback);
    }

    @Test(expected = IllegalArgumentException.class)
    public void getUserSegmentIdsNullSitegroups() {
        cxense.getUserSegmentIds(Collections.emptyList(), null, callback);
    }

    @Test
    public void getUser() {
        doNothing().when(cxense).getUser(any(UserIdentity.class), isNull(), isNull(), isNull(), eq(callback));
        cxense.getUser(mock(UserIdentity.class), callback);
        verify(cxense).getUser(any(UserIdentity.class), isNull(), isNull(), isNull(), eq(callback));
    }

    @Test
    public void getUserFullArgs() {
        cxense.getUser(mock(UserIdentity.class), Collections.emptyList(), false, Collections.emptyList(), callback);
        verifyApiCalled();
    }

    @Test(expected = IllegalArgumentException.class)
    public void getUserNullIdentity() {
        cxense.getUser(null, Collections.emptyList(), false, Collections.emptyList(), callback);
    }

    @Test
    public void getUserExternalData() {
        doNothing().when(cxense).getUserExternalData(anyString(), anyString(), eq(callback));
        cxense.getUserExternalData("type", callback);
        verify(cxense).getUserExternalData(null, "type", callback);
    }

    @Test
    public void getUserExternalDataFullArgs() {
        cxense.getUserExternalData("id", "type", callback);
        verifyApiCalled();
    }

    @Test
    public void setUserExternalData() {
        cxense.setUserExternalData(mock(UserExternalData.class), callback);
        verifyApiCalled();
    }

    @Test
    public void deleteUserExternalData() {
        cxense.deleteUserExternalData(mock(UserExternalData.class), callback);
        verifyApiCalled();
    }

    @Test
    public void getUserExternalLink() {
        cxense.getUserExternalLink("id", "type", callback);
        verifyApiCalled();
    }

    @Test
    public void setUserExternalLink() {
        cxense.setUserExternalLink("id", mock(UserIdentity.class), callback);
        verifyApiCalled();
    }

    @Test
    public void putEventTime() {
        doNothing().when(eventRepository).putEventTime(anyString(), anyLong());
        cxense.putEventTime("id", 99);
    }

    @Test
    public void pushEvents() {
        cxense.pushEvents(mock(Event.class));
        verify(executor).execute(any(Runnable.class));
    }

    @Test
    public void trackActiveTime() {
        doNothing().when(cxense).trackActiveTime(anyString(), anyLong());
        cxense.trackActiveTime("id");
        verify(cxense).trackActiveTime("id", 0);
    }

    @Test
    public void trackActiveTimeFullArgs() {
        cxense.trackActiveTime("id", 0);
        verify(executor).execute(any(Runnable.class));
    }

    @Test
    public void getDefaultUser() {
        cxense.getDefaultUser();
        verify(userProvider).getContentUser();
    }

    @Test
    public void flushEventQueue() {
        cxense.flushEventQueue();
        verify(sendTask).run();
    }

    @Test
    public void getQueueStatus() {
        when(eventRepository.getEventStatuses()).thenReturn(Collections.emptyList());
        QueueStatus status = cxense.getQueueStatus();
        assertNotNull(status);
        assertTrue(status.notSentEvents.isEmpty());
        assertTrue(status.sentEvents.isEmpty());
    }

    @Test
    public void setDispatchEventsCallback() {
        cxense.setDispatchEventsCallback(mock(DispatchEventsCallback.class));
        verify(sendTask).setDispatchEventsCallback(any(DispatchEventsCallback.class));
    }

    @Test
    public void executeGetPersistedQuery() {
        cxense.executePersistedQuery("url", "queryId", callback);
        verifyApiCalled();
    }

    @Test
    public void executePostPersistedQuery() {
        cxense.executePersistedQuery("url", "queryId", new Object(), callback);
        verifyApiCalled();
    }
}

File: sdk/src/test/java/com/cxense/cxensesdk/CxenseSdkTest.kt
package com.cxense.cxensesdk

import com.cxense.cxensesdk.model.ConsentSettings
import com.cxense.cxensesdk.model.UserIdentity
import com.cxense.cxensesdk.model.WidgetItem
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.doNothing
import com.nhaarman.mockitokotlin2.doReturn
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.never
import com.nhaarman.mockitokotlin2.spy
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import com.squareup.moshi.Moshi
import org.mockito.stubbing.Answer
import retrofit2.Call
import retrofit2.Callback
import java.util.concurrent.ScheduledExecutorService
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull

class CxenseSdkTest {
    private val call: Call<Any> = mock()

    private val executor: ScheduledExecutorService = mock()
    private val configuration: CxenseConfiguration = mock {
        on { consentSettings } doReturn ConsentSettings()
    }
    private val advertisingIdProvider: AdvertisingIdProvider = mock {
        on { defaultUserId } doReturn DEFAULT_USER_ID
        on { limitAdTrackingEnabled } doReturn false
    }
    private val userProvider: UserProvider = mock {
        on { userId } doReturn USER_ID
        on { defaultUser } doReturn mock()
    }
    private val cxApi: CxApi = mock<CxApi>(defaultAnswer = Answer { call })
    private val errorParser: ApiErrorParser = mock()
    private val moshi: Moshi = mock()
    private val eventRepository: EventRepository = mock()
    private val sendTask: SendTask = mock()

    private val cxenseSdk = spy(
        CxenseSdk(
            executor,
            configuration,
            advertisingIdProvider,
            userProvider,
            cxApi,
            errorParser,
            moshi,
            eventRepository,
            sendTask
        )
    )

    @Test
    fun getUserId() {
        assertEquals(USER_ID, cxenseSdk.userId)
        verify(userProvider).userId
    }

    @Test
    fun getDefaultUserId() {
        assertEquals(DEFAULT_USER_ID, cxenseSdk.defaultUserId)
        verify(advertisingIdProvider).defaultUserId
    }

    @Test
    fun getLimitAdTrackingEnabled() {
        assertFalse { cxenseSdk.limitAdTrackingEnabled }
        verify(advertisingIdProvider).limitAdTrackingEnabled
    }

    @Test
    fun setDispatchEventsCallback() {
        cxenseSdk.setDispatchEventsCallback(mock())
        verify(sendTask).sendCallback = any()
    }

    @Test
    fun pushEvents() {
        cxenseSdk.pushEvents(mock())
        verify(executor).execute(any())
    }

    @Test
    fun trackActiveTime() {
        cxenseSdk.trackActiveTime("")
        verify(executor).execute(any())
    }

    @Test
    fun flushEventQueue() {
        cxenseSdk.flushEventQueue()
        verify(executor).execute(eq(sendTask))
    }

    @Test
    fun getQueueStatus() {
        assertNotNull(cxenseSdk.queueStatus)
        verify(eventRepository).getEventStatuses()
    }

    @Test
    fun getDefaultContentUser() {
        assertNotNull(cxenseSdk.defaultContentUser)
        verify(userProvider).defaultUser
    }

    @Test
    fun trackClick() {
        val clickUrl = "clickUrl"
        doNothing().`when`(cxenseSdk).trackClick(any<String>(), any())
        cxenseSdk.trackClick(WidgetItem("title", "url", clickUrl, mapOf()), mock())
        verify(cxenseSdk).trackClick(eq(clickUrl), any())
    }

    @Test
    fun trackClickNullUrl() {
        val callback: LoadCallback<Unit> = mock()
        cxenseSdk.trackClick(WidgetItem("title", "url", null, mapOf()), callback)
        verify(callback).onError(any())
        verify(cxenseSdk, never()).trackClick(any<String>(), eq(callback))
    }

    @Test
    fun trackClickByUrl() {
        cxenseSdk.trackClick("url", mock())
        verify(call).enqueue(any<Callback<Any>>())
    }

    @Test
    fun loadWidgetRecommendations() {
        cxenseSdk.loadWidgetRecommendations("widgetId", callback = mock())
        verify(call).enqueue(any<Callback<Any>>())
        verify(cxenseSdk).defaultContentUser
    }

    @Test
    fun loadWidgetRecommendationsWithUser() {
        cxenseSdk.loadWidgetRecommendations("widgetId", user = mock(), callback = mock())
        verify(call).enqueue(any<Callback<Any>>())
        verify(cxenseSdk, never()).defaultContentUser
    }

    @Test
    fun reportWidgetVisibilities() {
        cxenseSdk.reportWidgetVisibilities(mock(), mock())
        verify(call).enqueue(any<Callback<Any>>())
    }

    @Test
    fun getUserSegmentIds() {
        cxenseSdk.getUserSegmentIds(listOf(mock()), listOf("123"), mock())
        verify(call).enqueue(any<Callback<Any>>())
    }

    @Test
    fun buildWithoutIdentities() {
        assertFailsWithMessage<IllegalArgumentException>("at least one user identity", "Expected fail for identities") {
            cxenseSdk.getUserSegmentIds(listOf(), listOf("sitegroupId"), mock())
        }
    }

    @Test
    fun buildWithoutSitegroups() {
        assertFailsWithMessage<IllegalArgumentException>(
            "at least one not empty site group id",
            "Expected fail for sitegroups"
        ) {
            cxenseSdk.getUserSegmentIds(listOf(mock()), listOf(""), mock())
        }
    }

    @Test
    fun getUserSegmentIdsNoConsent() {
        val callback: LoadCallback<List<String>> = mock()
        whenever(configuration.consentSettings).thenReturn(ConsentSettings().consentRequired(true))
        cxenseSdk.getUserSegmentIds(listOf(mock()), listOf("123"), callback)
        verify(callback).onError(any())
    }

    @Test
    fun getUser() {
        cxenseSdk.getUser(UserIdentity("cx", "id"), callback = mock())
        verify(call).enqueue(any<Callback<Any>>())
    }

    @Test
    fun getUserExternalData() {
        cxenseSdk.getUserExternalData("type", callback = mock())
        verify(call).enqueue(any<Callback<Any>>())
    }

    @Test
    fun setUserExternalData() {
        cxenseSdk.setUserExternalData(mock(), mock())
        verify(call).enqueue(any<Callback<Any>>())
    }

    @Test
    fun deleteUserExternalData() {
        cxenseSdk.deleteUserExternalData(mock(), mock())
        verify(call).enqueue(any<Callback<Any>>())
    }

    @Test
    fun getUserExternalLink() {
        cxenseSdk.getUserExternalLink("cxenseId", "type", mock())
        verify(call).enqueue(any<Callback<Any>>())
    }

    @Test
    fun addUserExternalLink() {
        cxenseSdk.addUserExternalLink("cxenseId", UserIdentity("cx", "id"), mock())
        verify(call).enqueue(any<Callback<Any>>())
    }

    @Test
    fun executePersistedQuery() {
        cxenseSdk.executePersistedQuery("url", "persistedQueryId", callback = mock<LoadCallback<Any>>())
        verify(call).enqueue(any<Callback<Any>>())
    }

    companion object {
        const val USER_ID = "user id"
        const val DEFAULT_USER_ID = "default user id"
    }
}
